beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -1849,13 +1861,123 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+            let mut potentially_abandoned_blocks: Vec<(
+                Slot,
+                Option<SignedBeaconBlockHash>,
+                Option<BeaconStateHash>,
+            )> = Vec::new();
+
+            for result in
+                SlotBlockStateIterator::from_block(Arc::clone(&self.store), head_hash.into())?
+            {
+                let (slot, is_skipped_slot, block_hash, state_hash) = result?;
+
+                if slot <= old_finalized_slot {
+                    // We must assume here any candidate chains include old_finalized_block_hash,
+                    // i.e. there aren't any forks starting at a block that is a strict ancestor of
+                    // old_finalized_block_hash.
+                    break;
+                }
+                if is_skipped_slot {
+                    potentially_abandoned_blocks.push((slot, None, Some(state_hash)));
+                } else {
+                    if newly_finalized_blocks.contains(&block_hash) {
+                        debug_assert!(slot <= new_finalized_slot);
+                        if slot == new_finalized_slot {
+                            // The first finalized block of a candidate chain lies after (in terms
+                            // of slots order) the newly finalized block.  It's not a candidate for
+                            // prunning.
+                            potentially_abandoned_blocks.clear();
+                            potentially_abandoned_head.take();
+                            break;
+                        }
+                    } else {
+                        potentially_abandoned_blocks.push((
+                            slot,
+                            Some(block_hash),
+                            Some(state_hash),
+                        ));
+                    }
+                }
+            }
+
+            if !potentially_abandoned_blocks.is_empty() {
+                let head_block = self
+                    .store
+                    .get_block(&head_hash)?
+                    .ok_or_else(|| BeaconStateError::MissingBeaconBlock(head_hash.into()))?;
+                potentially_abandoned_blocks.push((
+                    head_slot,
+                    Some(head_hash.into()),
+                    Some(head_block.state_root().into()),
+                ));
+
+                abandoned_blocks.extend(
+                    potentially_abandoned_blocks
+                        .iter()
+                        .filter_map(|(_, maybe_block_hash, _)| *maybe_block_hash),
+                );
+                abandoned_states.extend(potentially_abandoned_blocks.iter().filter_map(
+                    |(slot, _, maybe_state_hash)| match maybe_state_hash {
+                        None => None,
+                        Some(state_hash) => Some((*slot, *state_hash)),
+                    },
+                ));
+                abandoned_heads.extend(potentially_abandoned_head.into_iter());
```

Comment:
This could be `abandoned_heads.push(head_hash)`

Reply:
See my comment above.